### PR TITLE
Parser

### PR DESCRIPTION
**Pull request by passByReference**

### CODE REVIEW

File: parser/Parser.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
+from urllib import request, parse
+from html.parser import HTMLParser
+
+class AdItem(object):
+    def __init__(self, tag, pos, attr_name, value):
+        self._tag = tag
```
Comment (passByReference) [inline parser/Parser.py]: Will do

File: parser/Parser.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        self._html_page = data
+        print('\nTotal number of Ads found: {}'.format(len(self._ad_elements)))
+
+    def handle_starttag(self, tag, attrs):
+        '''
+        Scan all the start tag and find ad in the attributes
+        '''
+        for attr in attrs:
+            try:
+                if attr and 'ad' in attr[-1]:
```
Comment (passByReference) [inline parser/Parser.py]: Yes. This is a preliminary approach for ad-detecting. Placeholder for more advanced techniques.We can discuss over the weekend. I won't be working again on the weekend

File: parser/Parser.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            except:
+                print('Unexpected error')
+
+    def handle_startendtag(self, tag, attrs):
+        '''
+        Scans all the startend tag <xxx/> and find ad in the attributes
+        '''
+        for attr in attrs:
+            try:
+                if attr and 'ad' in attr[-1]:
```
Comment (passByReference) [inline parser/Parser.py]: These are inherited methods from HTTPParser which may not be moved out from parser. But it's a good point to have a utility component. I will think about it

File: parser/Parser.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
...252C32510%252C17281%252C17353%252C27815%252C2850%252C28478%252C17308%252C17305%252C17303%252C17309%252C17307%252C17306%252C17304%252C24653%252C24725%252C22980%252C17476%252C18890%252C17504%252C17468%252C18861%252C21865%252C22295%252C24077%252C24008%252C24009%252C24542%252C2988%252C27371%252C30352%252C8947%252C1913%252C30346%252C22061%252C3100%252C27087%252C8972%252C32497%252C27845%26amznslots%3Da3x2p12%252Ca7x9p12%252Ca3x6p10%252Ca9x2p12&amp;cookie=ID%3D17e3cd07072f7aa6%3AT%3D1489270186%3AS%3DALNI_MYDOC_N300ZbKBHsgEsrEtD3N1o_w&amp;abxe=1&amp;lmt=1494818471&amp;dt=1494818471229&amp;cc=100&amp;frm=20&amp;biw=1440&amp;bih=461&amp;oid=3&amp;adxs=900&amp;adys=725&amp;adks=3452988264&amp;gut=v2&amp;ifi=7&amp;u_tz=-240&amp;u_his=4&amp;u_h=900&amp;u_w=1440&amp;u_ah=823&amp;u_aw=1440&amp;u_cd=24&amp;u_nplug=4&amp;u_nmime=5&amp;u_sd=2&amp;flash=0&amp;url=https%3A%2F%2Fwww.theatlantic.com%2Fpolitics%2Farchive%2F2017%2F05%2Fif-there-are-white-house-recordings-they-could-be-subpoeanaed%2F526632%2F&amp;ref=https%3A%2F%2Fwww.theatlantic.com%2F&amp;dssz=102&amp;icsg=8590065664&amp;std=0&amp;vrg=117&amp;vrp=117&amp;ga_vid=1466052960.1489270178&amp;ga_sid=1494818337&amp;ga_hid=1128935240"></script>'
```
Comment (passByReference) [inline parser/Parser.py]: A test script. I will remove it later.

File: parser/Parser.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    p = Parser()
+    p.feed(res)
+    # print(p)
+
+    return
+
+main()
```
Comment (passByReference) [inline parser/Parser.py]: Ok thanks! Will update it in next commit

File: parser/Parser.py (hunk 2 above)
Comment (guanw) [inline parser/Parser.py]: LOL

**Pull request merged**